pkg/server/chaosd/kafka.go

```diff
@@ -0,0 +1,303 @@
+		if err != nil {
+			return nil, perr.Wrap(err, "create scram mechanism")
+		}
+	}
+	endpoint := fmt.Sprintf("%s:%d", attack.Host, attack.Port)
+	conn, err = dialer.DialLeader(context.Background(), "tcp", endpoint, attack.Topic, attack.Partition)
+	if err != nil {
+		return nil, perr.Wrapf(err, "dial kafka leader %s", endpoint)
+	}
+	return conn, nil
+}
+
+func attackKafkaFill(attack *core.KafkaCommand) (err error) {
+	// TODO: make it configurable
+	const messagePerRequest = 128
+	endpoint := fmt.Sprintf("%s:%d", attack.Host, attack.Port)
+	conn, err := client.DialLeader(context.Background(), "tcp", endpoint, attack.Topic, attack.Partition)
+	if err != nil {
+		return perr.Wrapf(err, "dial kafka leader %s", endpoint)
+	}
+	defer conn.Close()
+	msg := make([]byte, attack.MessageSize)
+	msgList := make([]client.Message, 0, messagePerRequest)
+	for i := 0; i < messagePerRequest; i++ {
+		msgList = append(msgList, client.Message{Value: msg})
+	}
+
+	counter := uint64(0)
+	start := time.Now()
+	written := "0 B"
+
+	for counter < attack.MaxBytes {
+		n, err := conn.WriteMessages(msgList...)
+		if err != nil {
+			return perr.Wrap(err, "write messages")
+		}
+		counter += uint64(n)
+		newWritten := humanize.Bytes(counter)
+		if newWritten != written {
+			written = newWritten
+			log.Info(fmt.Sprintf("write %s in %s", written, time.Now().Sub(start)))
+		}
+	}
+	return nil
+}
+
+func attackKafkaFlood(ctx context.Context, attack *core.KafkaCommand) (err error) {
+	endpoint := fmt.Sprintf("%s:%d", attack.Host, attack.Port)
+	conn, err := client.DialLeader(context.Background(), "tcp", endpoint, attack.Topic, attack.Partition)
+	if err != nil {
+		return perr.Wrapf(err, "dial kafka leader %s", endpoint)
+	}
+	defer conn.Close()
+	msg := make([]byte, attack.MessageSize)
+	timeout := time.Second / time.Duration(attack.RequestPerSecond)
+	wg := new(sync.WaitGroup)
+	for i := 0; i < int(attack.Threads); i++ {
+		logger := log.With(zap.String("thread", fmt.Sprintf("thread-%d", i)))
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			for {
+				select {
+				case <-ctx.Done():
+					return
+				default:
+					start := time.Now()
+					succeeded := 0
+					failed := 0
+					counter := uint64(0)
+					for time.Now().Sub(start) < time.Second && counter < attack.RequestPerSecond {
+						counter++
+						err := conn.SetWriteDeadline(time.Now().Add(timeout))
+						if err != nil {
+							failed++
+							if attack.NoSilent {
+								logger.Error("set write deadline", zap.Error(err))
+							}
+							continue
+						}
+						_, err = conn.Write(msg)
+						if err != nil {
+							failed++
+							if attack.NoSilent {
+								logger.Error("write message", zap.Error(err))
+							}
+							continue
+						}
+						succeeded++
+					}
+					logger.Info(fmt.Sprintf("succeeded: %d, failed: %d", succeeded, failed))
+					if time.Now().Sub(start) < time.Second {
+						time.Sleep(start.Add(time.Second).Sub(time.Now()))
```

Comment:
why need to sleep?

Reply:
To control the request-per-second. If `time.Now().Sub(start) < time.Second`, the target number of requests should be complete, so we should sleep for the next second.